crates/utils/src/tracing/grpc.rs

```diff
@@ -0,0 +1,116 @@
+use tracing_opentelemetry::OpenTelemetrySpanExt;
+
+/// A [`trace_fn`](tonic::transport::server::Server) implementation for the block producer which
+/// adds open-telemetry information to the span.
+///
+/// Creates an `info` span following the open-telemetry standard: `block-producer.rpc/{method}`.
+/// Additionally also pulls in remote tracing context which allows the server trace to be connected
+/// to the client's origin trace.
+pub fn block_producer_trace_fn(request: &http::Request<()>) -> tracing::Span {
+    let span = if let Some("SubmitProvenTransaction") = request.uri().path().rsplit('/').next() {
+        tracing::info_span!("block-producer.rpc/SubmitProvenTransaction")
+    } else {
+        tracing::info_span!("block-producer.rpc/Unknown")
+    };
+
+    add_otel_span_attributes(span, request)
+}
+
+/// A [`trace_fn`](tonic::transport::server::Server) implementation for the store which adds
+/// open-telemetry information to the span.
+///
+/// Creates an `info` span following the open-telemetry standard: `store.rpc/{method}`. Additionally
+/// also pulls in remote tracing context which allows the server trace to be connected to the
+/// client's origin trace.
+pub fn store_trace_fn(request: &http::Request<()>) -> tracing::Span {
+    let span = match request.uri().path().rsplit('/').next() {
+        Some("ApplyBlock") => tracing::info_span!("store.rpc/ApplyBlock"),
+        Some("CheckNullifiers") => tracing::info_span!("store.rpc/CheckNullifiers"),
+        Some("CheckNullifiersByPrefix") => tracing::info_span!("store.rpc/CheckNullifiersByPrefix"),
+        Some("GetAccountDetails") => tracing::info_span!("store.rpc/GetAccountDetails"),
+        Some("GetAccountProofs") => tracing::info_span!("store.rpc/GetAccountProofs"),
+        Some("GetAccountStateDelta") => tracing::info_span!("store.rpc/GetAccountStateDelta"),
+        Some("GetBlockByNumber") => tracing::info_span!("store.rpc/GetBlockByNumber"),
+        Some("GetBlockHeaderByNumber") => tracing::info_span!("store.rpc/GetBlockHeaderByNumber"),
+        Some("GetBlockInputs") => tracing::info_span!("store.rpc/GetBlockInputs"),
+        Some("GetBatchInputs") => tracing::info_span!("store.rpc/GetBatchInputs"),
+        Some("GetNotesById") => tracing::info_span!("store.rpc/GetNotesById"),
+        Some("GetTransactionInputs") => tracing::info_span!("store.rpc/GetTransactionInputs"),
+        Some("SyncNotes") => tracing::info_span!("store.rpc/SyncNotes"),
+        Some("SyncState") => tracing::info_span!("store.rpc/SyncState"),
+        _ => tracing::info_span!("store.rpc/Unknown"),
+    };
```

Comment:
`tracing::span!` requires a compile-time constant string.

A more maintenance free alternative is to allow the open-telemetry and stdout tracing to drift apart some more. We could set the `otel.name` property which overrides the span name for open-telemetry (currently it just defaults to the span name).

This would mean stdout would get a more generic `name = store.rpc` and a separate `rpc.method = {method}` attribute.

Might be fine/better than this manual matching?

Reply:
Ah, I think this addresses the m question above. I think we can leave it for a future PR.